### DISCUSSION TITLE
Is there a way to `fail` a template silently?

### DISCUSSION THREAD

**Discussion by felipecrs** [twpayne/chezmoi#1670] [Q&A] [answered]
I'm doing this:

```
{{- $chezmoiForce := or (has "--force" .chezmoi.args) (has "--force=true" .chezmoi.args) -}}

{{ $name := "" -}}
{{- if hasKey . "name" -}}
{{-   $name = .name -}}
{{- end -}}
{{- if and stdinIsATTY (not $chezmoiForce) -}}
{{-   $question := "What is your full name" -}}
{{-   if $name -}}
{{-     $name = promptString $question $name -}}
{{-   else -}}
{{-     $name = promptString $question -}}
{{-   end -}}
{{-   if $name -}}
{{-     writeToStdout (printf "✔ Name set as '%s'\n" $name) -}}        
{{-   else -}}
{{-     writeToStdout "❌ Invalid name, please start over.\n" -}}
{{-     fail "" -}}
{{-   end -}}
{{- end -}}
```

But it fails like:

```console
❯ chezmoi init --data=false
What is your full name?
❌ Invalid name, please start over.
chezmoi: template: chezmoi.yaml:18:8: executing "chezmoi.yaml" at <fail "">: error calling fail:
```

It would be cool if I could simply fail with a stdout message, rather than a "debug" mesage.

**Comment by twpayne** [comment]
I don't think this is possible. As mentioned in other discussions and issues, `text/template` is a templating language designed to render static data, not a programming language.

**Reply by felipecrs** [reply]
> I don't think this is possible. 

May I suggest a new initTemplateFunc called `failWithStdout *message*` or just `failSilently`?

> text/template is a templating language designed to render static data, not a programming language.

I don't see how this is relevant here.

**Reply by twpayne** [reply]
> May I suggest a new initTemplateFunc called `failWithStdout *message*` or just `failSilently`?

I don't think this is possible. [`text/template.Template.Execute`](https://pkg.go.dev/text/template#Template.Execute) either succeeds (in which case it returns `nil`) or fails (in which case it returns an error). There is no concept of early termination.

> > text/template is a templating language designed to render static data, not a programming language.
> 
> I don't see how this is relevant here.

It's relevant because programming languages have richer control flow, e.g. while loops, returns, etc. The template in the original post is not rendering static data, it's trying to run an interactive loop.

**Reply by felipecrs** [reply]
> I don't think this is possible. [`text/template.Template.Execute`](https://pkg.go.dev/text/template#Template.Execute) either succeeds (in which case it returns `nil`) or fails (in which case it returns an error). There is no concept of early termination.

I think it's possible by returning a specific error from the new function and treating it before exiting chezmoi to handle the exit message. No?

> It's relevant because programming languages have richer control flow, e.g. while loops, returns, etc. The template in the original post is not rendering static data, it's trying to run an interactive loop.

The loop itself is being discussed at https://github.com/twpayne/chezmoi/discussions/1671. There is no loop here.

Regardless, sprig provides the `fail` function [just because of it: flow control](https://masterminds.github.io/sprig/flow_control.html). I'm not reinventing the wheel as sprig already provides it, I just want to improve the message.

**Reply by twpayne** [reply]
> I think it's possible by returning a specific error from the new function and treating it before exiting chezmoi to handle the exit message. No?

You're right. #1674 should implement this.

**Reply by twpayne** [reply]
With #1674, failing silently can be achieved with `{{ exit 0 }}`. chezmoi must be built with Go 1.17 or later for this to work.

**Comment by twpayne** [comment, marked as answer]
With #1674, chezmoi now has an `exit` template function which allows you to fail templates silently with `{{ exit 0 }}`. This function is only available when generating the initial config file (#1771).
